File: solutions/clojure/list-ops/1/src/list_ops.clj
(ns list-ops)

(defn append 
  "Given two vectors, it adds all the items in the second vector to the end of the first vector"
  [coll1 coll2]
  (reduce conj coll1 coll2))

(defn concatenate 
  "Given a vector of vectors, it combines all the vectors into one flattened vector"
  [colls]
  (reduce append [] colls))

(defn select-if
  "Given a predicate and a vector, it returns the vector of all items for which predicate(item) is true"
  [pred coll]
  (filter pred coll))

(defn length 
  "Given a vector, it returns the number of items within it"
  [coll]
  (count coll))

(defn apply-to-each 
  "Given a function and a vector, it returns the vector of the results of applying function(item) on all items"
  [f coll]
  (map f coll))

(defn foldl 
  "Given a function, a vector, and initial accumulator, it folds (reduces) each item into the accumulator from the left"
  [f coll acc]
  (reduce f acc coll))

(defn foldr
  "Given a function, a vector, and an initial accumulator, it folds (reduces) each item into the accumulator from the right"
  [f coll acc]
  (->> coll
    reverse
    (reduce f acc)))

(defn reverse-order 
  "Given a vector, it returns a vector with all the original items, but in reverse order"
  [coll]
  (reverse coll))

File: solutions/go/list-ops/1/list_ops.go
package listops

// IntList is an abstraction of a list of integers which we can define methods on
type IntList []int

func (s IntList) Foldl(fn func(int, int) int, initial int) int {
	for _, n := range s {
		initial = fn(initial, n)
	}
	return initial
}

func (s IntList) Foldr(fn func(int, int) int, initial int) int {
	for i := len(s) - 1; i >= 0; i-- {
		initial = fn(s[i], initial)
	}
	return initial
}

func (s IntList) Filter(fn func(int) bool) IntList {
	filtered := []int{}
	for _, n := range s {
		if fn(n) {
			filtered = append(filtered, n)
		}
	}
	return filtered
}

func (s IntList) Length() int {
	return len(s)
}

func (s IntList) Map(fn func(int) int) IntList {
	mapped := []int{}
	for _, n := range s {
		mapped = append(mapped, fn(n))
	}
	return mapped
}

func (s IntList) Reverse() IntList {
	reversed := []int{}
	for i := len(s) - 1; i >= 0; i-- {
		reversed = append(reversed, s[i])
	}
	return reversed
}

func (s IntList) Append(lst IntList) IntList {
	return append(s, lst...)
}

func (s IntList) Concat(lists []IntList) IntList {
	l := append([]int{}, s...)
	for _, xs := range lists {
		l = append(l, xs...)
	}
	return l
}
